presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationContext.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.verifier.framework;
+
+import com.facebook.presto.jdbc.QueryStats;
+import com.facebook.presto.verifier.event.FailureInfo;
+import com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup;
+
+import java.util.EnumMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup.CONTROL;
+import static com.facebook.presto.verifier.framework.QueryOrigin.QueryGroup.TEST;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class VerificationContext
+{
+    private Map<QueryGroup, Set<QueryException>> failures = new EnumMap<>(QueryGroup.class);
+
+    public VerificationContext()
+    {
+        failures.put(CONTROL, new LinkedHashSet<>());
+        failures.put(TEST, new LinkedHashSet<>());
+    }
+
+    public void recordFailure(QueryException exception)
+    {
+        failures.get(exception.getQueryOrigin().getGroup()).add(exception);
+    }
+
+    public List<FailureInfo> getAllFailures(QueryGroup group)
+    {
+        return failures.get(group).stream()
+                .map(exception -> new FailureInfo(
+                        exception.getQueryOrigin().getStage(),
+                        exception.getErrorCode(),
```

Comment:
Looking at how `QueryException` is generated for Cluster connection issues:

```
    public static QueryException forClusterConnection(Throwable cause)
    {
        return new QueryException(cause, CLUSTER_CONNECTION, Optional.empty(), true, Optional.empty());
    }

```
the error code is empty. So in this situation, for cluster connection issues - we will essentially pass in an empty error code and there will be no information about what happened. Should we not just serialize and return the whole `QueryException` object?

Reply:
- `getErrorCode` is different from `getPrestoErrorCode` - it is always a non-empty string. It is a short description of the error. Examples are `PRESTO(PERMISSION_ERROR)` or `CLUSTER_CONNECTION(SocketTimeoutException)`

Reply:
Thanks for the clarification on the `getErrorCode` method - I missed that.

Re: stacktrace: atleast based on my previous work in Hive - stack traces have been an integral part of debugging. I don't have an exact example that I can cite as of now to justify it, but overall, I can't imagine digging into specifics without stack traces. If you just give me `SocketTimeoutException` that gives me very few details and pointers to start debugging.

Since you have been resistant to add it, I am curious to hear your point of view of not adding stack traces. If the only reason is that you feel it is not useful, I will differ in opinion here. Is there some other reason as well?